rust/src/detect/transform_base64.rs

```diff
@@ -0,0 +1,420 @@
+            flags: 0,
+            nbytes: 0,
+            nbytes_str: std::ptr::null_mut(),
+            offset: 0,
+            offset_str: std::ptr::null_mut(),
+            mode: TRANSFORM_FROM_BASE64_MODE_DEFAULT,
+        }
+    }
+}
+
+impl DetectTransformFromBase64Data {
+    pub fn new() -> Self {
+        Self {
+            ..Default::default()
+        }
+    }
+}
+
+fn get_mode_value(value: &str) -> Result<DetectBase64Mode, ()> {
+    let res = match value {
+        "rfc4648" => DetectBase64Mode::ModeRFC4648,
+        "rfc2045" => DetectBase64Mode::ModeRFC2045,
+        "strict" => DetectBase64Mode::ModeStrict,
+        _ => return Err(()),
+    };
+
+    Ok(res)
+}
+
+fn parse_transform_base64(
+    input: &str,
+) -> IResult<&str, DetectTransformFromBase64Data, RuleParseError<&str>> {
+    // Inner utility function for easy error creation.
+    fn make_error(reason: String) -> nom7::Err<RuleParseError<&'static str>> {
+        Err::Error(RuleParseError::InvalidTransformBase64(reason))
+    }
+    let mut transform_base64 = DetectTransformFromBase64Data::new();
+
+    if input.is_empty() {
+        return Ok((input, transform_base64));
+    }
+    let (_, values) = nom7::multi::separated_list1(
+        tag(","),
+        preceded(multispace0, nom7::bytes::complete::is_not(",")),
+    )(input)?;
+
+    // No options so return defaults
+    if values.is_empty() {
+        return Ok((input, transform_base64));
+    };
+
+    if values.len() < DETECT_TRANSFORM_BASE64_MIN_PARAM_COUNT
+        || values.len() > DETECT_TRANSFORM_BASE64_MAX_PARAM_COUNT
+    {
+        return Err(make_error(format!("Incorrect argument string; at least {} values must be specified but no more than {}: {:?}",
+            DETECT_TRANSFORM_BASE64_MIN_PARAM_COUNT, DETECT_TRANSFORM_BASE64_MAX_PARAM_COUNT, input)));
+    }
+
+    for value in values {
+        let (mut val, mut name) = take_until_whitespace(value)?;
+        val = val.trim();
+        name = name.trim();
+        match name {
+            "mode" => {
+                if 0 != (transform_base64.flags & DETECT_TRANSFORM_BASE64_FLAG_MODE) {
+                    return Err(make_error("mode already set".to_string()));
+                }
+                transform_base64.mode = match get_mode_value(val) {
+                    Ok(val) => val,
+                    Err(_) => {
+                        return Err(make_error(format!("invalid mode value: {}", val)));
+                    }
+                };
+                transform_base64.flags |= DETECT_TRANSFORM_BASE64_FLAG_MODE;
+            }
+
+            "offset" => {
+                if 0 != (transform_base64.flags & DETECT_TRANSFORM_BASE64_FLAG_OFFSET) {
+                    return Err(make_error("offset already set".to_string()));
+                }
+
+                let (_, res) = parse_var(val)?;
+                match res {
+                    ResultValue::Numeric(val) => {
+                        if val as u32 <= 65535 {
```

Comment:
Why should we need `as u32` ?

Reply:
To detect out of range values, e.g., `70000`

Reply:
But `val` is u64 so you can just have `if val <= 65535 {` and that will work for `70000` (and also for `UINT32_MAX+1`)